source/common/config/grpc_mux_impl.cc

```diff
@@ -228,11 +232,30 @@ void GrpcMuxImpl::onStreamEstablished() {
   }
 }
 
-void GrpcMuxImpl::onEstablishmentFailure() {
-  for (const auto& api_state : api_state_) {
-    for (auto watch : api_state.second.watches_) {
-      watch->callbacks_.onConfigUpdateFailed(
-          Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure, nullptr);
+void GrpcMuxImpl::onEstablishmentFailure(bool remote_close) {
+  for (auto& api_state : api_state_) {
+    // check whether attempted type url is alpha version
```

Comment:
So, is fallback happening on resource API version? Isn't transport API fallback what we really want? I imagine for resource, it's plausible a management server just throws its hands up when asked for a v3 resource and returns the v2 resource, Envoy can still process this transparently via its config upgrade at ingestion.

Reply:
@htuch I didn't pushed anyway, There is resource level fallback. In this case, we can retry with updating `type_url` and disposing current watches if response has `NotFound`.  So, current implementation is incomplete on this PR. I will push them if I finished to write delta gRPC unit testing.

Reply:
I'm not sure how useful resource vs. transport API fallback is here.. can you articulate some likely scenarios where it will be useful?

Reply:
@htuch if management server passed older version configuration, can envoy process all of them? If it includes deleted fields, how it process? In addition, do you suppose transport fallback specifically? Would you mind telling me about your idea about them?

Reply:
So, I think the idea is that there would be a mode where Envoy would first attempt to reach out to a management server on the v3 endpoint. If this fails, e.g. due to 404 equivalent, it falls back to the v2 endpoint. I'm not convinced this is necessary given we have support to specify version in `ConfigSource` but would be in the nice-to-have category.

Reply:
@htuch recreated https://github.com/envoyproxy/envoy/pull/9799.
For now, I implemented gRPC fallback only. Is it conforming to your requirements?